### PR TITLE
Broadcast masked_select and implement double backwards for masked_scatter, index_copy, scatter_add, gather

### PR DESCRIPTION
**Pull request by gchanan**
1) Add broadcasting to masked_select; it was arbitrary that we weren't broadcasting it previously (there is no exact numpy equivalent outside of advanced indexing).  For consistency with similar functions, it's probably better to broadcast it.
2) Implement double backwards of masked_scatter; this is easier with masked_select broadcasting because those two functions are the forward/backwards inverses.

### CODE REVIEW

**Comment by soumith** [comment]
IndexCopy test failed in the travis once.

**Comment by gchanan** [comment]
this has another occasional failure: if the generated mask for mask_select is all 0s, the test will fail because the output is 0-dimensional and various places in the code can't handle that.  This would previously happen once every 2^25 times (because the mask was 5x5 in the smallest case), but now the smallest mask only is generated with 5 elements and expanded, so it fails every 2^5 times.

**Comment by gchanan** [comment]
The problem with indexCopy is the output and gradient are undefined in the presence of repeated indices.  The docs are incorrect, they say that indices is "index (LongTensor) – Indices to select from tensor", but it's actually indices to copy to the underlying tensor, so with repeated indices its undefined who "wins".

Here's an example with cuda:
```
>>> torch.zeros(1000,1000).cuda().index_copy_(0, torch.zeros(1000).long().cuda(), torch.arange(1,1001).expand(1000,1000).t().cuda())

  987   987   987  ...    984   984   984
    0     0     0  ...      0     0     0
    0     0     0  ...      0     0     0
       ...          ⋱          ...
    0     0     0  ...      0     0     0
    0     0     0  ...      0     0     0
    0     0     0  ...      0     0     0
[torch.cuda.FloatTensor of size 1000x1000 (GPU 0)]
```

**Comment by gchanan** [comment]
I believe the above issues have been addressed.

**Review by soumith — APPROVED** [review]
